Support >16 byte struct alignment, add double vec/matrix types

LPGhatguy opened the pull request:
Closes #4.

This PR does a couple things that are intertwined. I'll try to explain each thing and why it's tangled with the others!

First, I added support for structs with alignment over 16 bytes. Thanks to @benmkw in [this comment](https://github.com/LPGhatguy/crevice/issues/1#issuecomment-711087465) for pointing out this need!

In order to write sufficient tests for that increased alignment, I needed base types with alignment greater than 16. Instead of adding fake types, I added the double types that we already wanted anyways. I'm not 100% sure about the layout of the double-based matrix types and would love some extra review.

In order to add the new vector and matrix types, I decided it was probably time to switch the vector and matrix declarations to use a macro. I figure this will be useful as part of #3 as well in the future, as well as if/when we introduce the int and bool-based vectors and matrices.

benmkw commented on src/std140/primitives.rs:
```diff
@@ -18,101 +18,119 @@ unsafe impl Std140 for u32 {
+    Mat2 {
+        x: Vec2,
+        _pad_y: [f32; 2],
+        y: Vec2,
```
shouldn't there be padding here as well?

benmkw replied:
thinking about it: since vec2s other than dvec2 are align 8, maybe there should be no padding at all?

LPGhatguy replied:
The way that alignment works in Crevice is kind of strange! For manually created std140 structs like this one, the padding needs to be explicitly declared because these types don't have extra alignment according to rustc (ie they have no `#[repr(align)]`)

I think you're right that `Mat2` should have trailing padding here, and `Mat3` should as well if it doesn't already.

benmkw commented on src/std140/primitives.rs:
```diff
@@ -18,101 +18,119 @@ unsafe impl Std140 for u32 {
+}
+
+matrices! {
+    #[doc = "Corresponds to GLSL's `mat2`."]
```
should these maybe read `GLSL with std140` instead of `GLSL`?

benmkw commented:
we had the following an example at university which could maybe be added as a test case

<img width="503" alt="Bildschirmfoto 2020-10-18 um 11 17 31" src="https://user-images.githubusercontent.com/57018992/96363324-d0dbc480-1133-11eb-9c0d-a4995e6119dc.png">

benmkw commented:
Another test case could be using e.g. wgpu compute https://github.com/gfx-rs/wgpu-rs/blob/master/examples/hello-compute/main.rs with several different nested layouts and checking if the results match up as expected.

benmkw commented on crevice-derive/src/lib.rs:
```diff
@@ -75,6 +75,16 @@ pub fn derive_as_std140(input: CompilerTokenStream) -> CompilerTokenStream {
+    let struct_alignment = fields.named.iter().fold(quote!(16usize), |last, field| {
+        let field_ty = &field.ty;
+        quote! {
+            ::crevice::internal::max(
```
I think some comment here would be good to highlight this part as being important

LPGhatguy replied:
I'll do you one better and break this out into a new function. We'll need to eventually, and this whole proc macro needs a little love honestly.

benmkw commented:
I may have missed it but there does not seem to be support for arrays with members with alignment of more than 16? Or did this work before correctly already? Cause I see the `max` operation only being done for structs.

LPGhatguy commented:
I don't think Crevice supports arrays yet! That's still on hold with #5.

LPGhatguy commented:
I've added trailing padding to matrices (which should line up with the spec), updated comments, and made the proc macro struct alignment generation a little easier to grok.

LPGhatguy commented:
Going to go ahead and merge this one, we can follow up with more changes (and more releases!) if I made any mistakes here.

benmkw commented:
so this looks correct to me, trying to do the example from above
(I think it would be good if the types would derive Default btw.)
```
#[derive(AsStd140, Default, TypeLayout)]
struct AnyBlock {
    a: f32,
    b: Anon,
    c: f32,
    // d: crate::Vec3, // or TODO array
    // e: cgmath::Matrix3<f32>,
    f: f32,
}

#[derive(AsStd140, Default, TypeLayout)]
struct Anon {
    u: f32,
    v: f32,
}


Std140AnyBlock (size 32, alignment 4)
| Offset | Name     | Size |
| ------ | -------- | ---- |
| 0      | a        | 4    |
| 4      | _b_align | 12   |
| 16     | b        | 8    |
| 24     | c        | 4    |
| 28     | f        | 4    |
```

<details>


```diff
diff --git a/src/std140/primitives.rs b/src/std140/primitives.rs
index b2f6665..0aa2ead 100644
--- a/src/std140/primitives.rs
+++ b/src/std140/primitives.rs
@@ -27,7 +27,7 @@ macro_rules! vectors {
         $(
             #[$doc]
             #[allow(missing_docs)]
-            #[derive(Debug, Clone, Copy)]
+            #[derive(Debug, Clone, Copy, Default)]
             pub struct $name {
                 $(pub $field: $prim,)+
             }
diff --git a/tests/std140.rs b/tests/std140.rs
index 112849a..79bec9e 100644
--- a/tests/std140.rs
+++ b/tests/std140.rs
@@ -3,6 +3,36 @@ use type_layout::TypeLayout;
 
 use crevice::std140::{AsStd140, DVec4, Std140, Vec3};
 
+#[derive(AsStd140, Default, TypeLayout)]
+struct AnyBlock {
+    a: f32,
+    b: Anon,
+    c: f32,
+    // d: crate::Vec3,
+    // e: cgmath::Matrix3<f32>,
+    f: f32,
+}
+
+#[derive(AsStd140, Default, TypeLayout)]
+struct Anon {
+    u: f32,
+    v: f32,
+}
+
+#[test]
+fn anyBlock() {
+    println!("{}", <AnyBlock as AsStd140>::Std140Type::type_layout());
+    assert_eq!(<AnyBlock as AsStd140>::Std140Type::ALIGNMENT, 16); 
+    assert_eq!(
+        std::mem::size_of::<<AnyBlock as AsStd140>::Std140Type>(),
+        49 // wrong on purpose to make cargo test show output
+    );
+
+    let value = AnyBlock::default();
+
+    let _value_std140 = value.as_std140();
+}
+
 #[derive(AsStd140)]
 struct PrimitiveF32 {
     x: f32,
```
</details>


for 

```
#[derive(AsStd140, Default, TypeLayout)]
struct AnyBlock {
    a: f32,
    b: Anon,
    c: f32,
    d: crate::Vec3,
    // e: cgmath::Matrix3<f32>,
    f: f32,
}

#[derive(AsStd140, Default, TypeLayout)]
struct Anon {
    u: f32,
    v: f32,
}


| Offset | Name     | Size |
| ------ | -------- | ---- |
| 0      | a        | 4    |
| 4      | _b_align | 12   |
| 16     | b        | 8    |
| 24     | c        | 4    |
| 28     | _d_align | 4    |
| 32     | d        | 12   |
| 44     | f        | 4    |
```
I'm wondering though, the padding for d should really be after d not before d?

benmkw commented:
It also does not seem possible to have cgmath::Vec3<f32> as a member? (tried that instead of mint cause they derive Default which makes testing more concise)

LPGhatguy commented:
> (I think it would be good if the types would derive Default btw.)

Generally, the actual underlying `std140` types aren't intended to be used as-is. If you really want to, you can use bytemuck's `Zeroed::zeroed()` instead!

> It also does not seem possible to have cgmath::Vec3 as a member? (tried that instead of mint cause they derive Default which makes testing more concise)

It's actually very tricky to get Crevice to support any math library's types! I filed [this issue with mint](https://github.com/kvark/mint/issues/59) to add a trait that would help do this.

benmkw commented:
very interesting issue, thanks for sharing!
